peps/pep-0734.rst

```diff
@@ -0,0 +1,400 @@
+       interp.exec(tw.dedent("""
+           def handle_request(req):
+               ...
+
+           def capture_exception(exc):
+               ...
+
+           while True:
+               try:
+                   req = tasks.recv()
+               except Exception:
+                   # channel closed
+                   break
+               try:
+                   res = handle_request(req)
+               except Exception as exc:
+                   res = capture_exception(exc)
+               results.send_nowait(res)
+           """))
+   threads = [threading.Thread(target=worker) for _ in range(20)]
+   for t in threads:
+       t.start()
+
+   requests = ...
+   for req in requests:
+       tasks.send(req)
+   tasks.close()
+
+   for t in threads:
+       t.join()
+
+Sharing a memoryview (imagine map-reduce):
+
+::
+
+   data, chunksize = read_large_data_set()
+   buf = memoryview(data)
+   numchunks = (len(buf) + 1) / chunksize
+   results = memoryview(b'\0' * numchunks)
+
+   tasks_recv, tasks = interpreters.create_channel()
+
+   def worker():
+       interp = interpreters.create()
+       interp.set_main_attrs(data=buf, results=results, tasks=tasks_recv)
+       interp.exec(tw.dedent("""
+           while True:
+               try:
+                   req = tasks.recv()
+               except Exception:
+                   # channel closed
+                   break
+               resindex, start, end = req
+               chunk = data[start: end]
+               res = reduce_chunk(chunk)
+               results[resindex] = res
+           """))
+   t = threading.Thread(target=worker)
+   t.start()
+
+   for i in range(numchunks):
+       if not workers_running():
+           raise ...
+       start = i * chunksize
+       end = start + chunksize
+       if end > len(buf):
+           end = len(buf)
+       tasks.send((start, end, i))
+   tasks.close()
+   t.join()
+
+   use_results(results)
+
+
+Documentation
+=============
+
+The new stdlib docs page for the ``interpreters`` module will include
+the following:
+
+* (at the top) a clear note that support for multiple interpreters
+  is not required from extension modules
+* some explanation about what subinterpreters are
+* brief examples of how to use multiple interpreters
+  (and communicating between them)
+* a summary of the limitations of using multiple interpreters
+* (for extension maintainers) a link to the resources for ensuring
+  multiple interpreters compatibility
+* much of the API information in this PEP
+
+Docs about resources for extension maintainers already exist on the
+`Isolating Extension Modules <isolation-howto_>`_ howto page.  Any
+extra help will be added there.  For example, it may prove helpful
+to discuss strategies for dealing with linked libraries that keep
+their own subinterpreter-incompatible global state.
+
+.. _isolation-howto:
+   https://docs.python.org/3/howto/isolating-extensions.html
```

Comment:
We can link using Intersphinx:
```suggestion
:ref:`python:isolating-extensions-howto` howto page.  Any
extra help will be added there.  For example, it may prove helpful
to discuss strategies for dealing with linked libraries that keep
their own subinterpreter-incompatible global state.
```